[JSC] Various improvements to the aboutToMarkSlow instrumentation

dhecht opened the pull request:
#### 3e43b2b32ed3788fd6c0b16a527373ca8b890edc
<pre>
[JSC] Various improvements to the aboutToMarkSlow instrumentation
<a href="https://bugs.webkit.org/show_bug.cgi?id=283176">https://bugs.webkit.org/show_bug.cgi?id=283176</a>
<a href="https://rdar.apple.com/139971430">rdar://139971430</a>

Reviewed by Mark Lam.

The MarkedBlock::Header::m_lock field sits at offset 24.
dumpInfoAndCrashForInvalidHandleV2() is always called with this
lock held, and so regardless of whether &apos;m_lock&apos; was corrupted with
a zero, it will always be non-zero by the time the instrumentation reads it.
So that we can get a better guess at how many contiguious bytes
from the start of the MarkedBlock are corrupted with zeros, treat the bytes
occupied by m_lock as zeros when scanning for zeros.

In other words, we don&apos;t know what value these bytes were corrupted with,
so assume they were corrupted with zero so that the scan for zeros continues.

Also count the total number of zeros in the MarkedBlock (which is
the same as a page on Darwin) to get an overall view into the state
of the page, to account for the fact that there could be other
locations on the page that may have been written after the corruption
occurs.

Improve the use of WTF::setCrashLogMessage(). That function saves
only the last message, so send the same message on each invocation
but with additional known details filled in. Then, include
the source code line number so that the message can be interpreted
properly.

Record the first 8 bytes of the HeapCell rather than just the type.

Remove the code that tries to save values to registers in the middle
of the instrumentation. It doesn&apos;t work in practice since a crash
would be more likely to occur after running non-trivial code, which
will likely clobber the registers anyway.

Add some testing code for the various instrumentation paths
(disabled at compile time).

* Source/JavaScriptCore/heap/MarkedBlock.cpp:
(JSC::MarkedBlock::setupTestForDumpInfoAndCrash):
(JSC::MarkedBlock::aboutToMarkSlow):
(JSC::MarkedBlock::dumpInfoAndCrashForInvalidHandleV2):
(JSC::MarkedBlock::dumpInfoAndCrashForInvalidHandle): Deleted.
* Source/JavaScriptCore/heap/MarkedBlock.h:
(JSC::MarkedBlock::setupTestForDumpInfoAndCrash):
* Source/JavaScriptCore/runtime/OptionsList.h:

Canonical link: <a href="https://commits.webkit.org/286689@main">https://commits.webkit.org/286689@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c5f86c95531f0d5a24ddb16592cd04c9179eb300

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76703 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55738 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29609 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81235 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27979 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64880 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4031 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60129 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18207 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79770 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50054 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65856 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40451 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47454 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23350 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-masking/clip-path/animations/clip-path-animation-font-size-mixed-change.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26303 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/69886 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68574 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23680 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82680 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/75980 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4079 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2694 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68405 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4232 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65828 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67657 "Found 1 new API test failure: /TestWebKit:WebKit.InjectedBundleInitializationUserDataCallbackWins (failure)") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/16882 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11627 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9706 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/98234 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11879 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4026 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21488 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4049 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7479 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5807 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dhecht commented on Source/JavaScriptCore/heap/MarkedBlock.cpp:
```diff
@@ -210,17 +210,57 @@ void MarkedBlock::Handle::resumeAllocating(FreeList& freeList)
+            memset(blockMem + 3 * blockSize / 4, 0, blockSize / 4);
+            break;
+        case 4: // Test contiguous and total zero byte counts: entire block zeroed.
+            dataLogLn("Zeroing start and end of MarkedBlock");
```
Oops will fix this print msg

MenloDorian commented on Source/JavaScriptCore/heap/MarkedBlock.cpp:
```diff
@@ -210,17 +210,57 @@ void MarkedBlock::Handle::resumeAllocating(FreeList& freeList)
     sweep(&freeList);
 }
 
+#if ENABLE(MARKEDBLOCK_TEST_DUMP_INFO)
```
typo in the commit msg: /invokation/invocation/ and /interpretted/interpreted/.

Please also add a comment explaining why `m_lock` "will always be non-zero" so that it's understood the code path taken to reach the diagnostics will acquire the lock thereby setting it.  It's been set is not an indication that the block hasn't been inadvertently cleared by some external agent.

dhecht replied:
> Please also add a comment explaining why m_lock "will always be non-zero"

There is a code comment already explaining m_lock, but I'll enhance the commit message, assuming that's what you mean.

> It's been set is not an indication that the block hasn't been inadvertently cleared by some external agent.

Of course. I'm not sure which part of the commit message implies this.

Oh, maybe this, which I can clarify:

> locations on the page that are written before corruption is detected.

I'll say instead: "... may have been written after the corruption occurs."

MenloDorian commented on Source/JavaScriptCore/heap/MarkedBlock.cpp:
```diff
@@ -513,156 +553,105 @@ void MarkedBlock::Handle::sweep(FreeList* freeList)
+        WTF::setCrashLogMessage(msg);
+        dataLogLn(msg);
+#else
+    UNUSED_PARAM(line);
```
indent by 4 please.

MenloDorian commented on Source/JavaScriptCore/heap/MarkedBlock.h:
```diff
@@ -425,8 +425,9 @@ class MarkedBlock {
 
-    // This is only used for debugging. We should remove this once the issue is resolved (rdar://136782494)
-    NO_RETURN_DUE_TO_CRASH NEVER_INLINE void dumpInfoAndCrashForInvalidHandle(AbstractLocker&, HeapCell*);
+    // FIXME: remove these rdar://136782494) is resolved
```
Please replace this comment with a simple `FIXME: rdar://.....` where the radar is an unresolved bug and tracks the work to clean this up.  That way, when one looks up the radar in the future, it will tell us what to clean up.

dhecht replied:
Okay. I was just simplifying the existing comment, but I'm happy to create a new radar for this follow on cleanup.

MenloDorian commented on Source/JavaScriptCore/runtime/OptionsList.h:
```diff
@@ -604,6 +604,7 @@ bool hasCapacityToUseLargeGigacage();
     v(Bool, useWasmSIMD, true, Normal, "Allow the new simd instructions and types from the wasm simd spec."_s) \
     v(Bool, useWasmRelaxedSIMD, false, Normal, "Allow the relaxed simd instructions and types from the wasm relaxed simd spec."_s) \
     v(Bool, useWasmTailCalls, true, Normal, "Allow the new instructions from the wasm tail calls spec."_s) \
+    v(Unsigned, markedBlockDumpInfoCount, 0, Normal, nullptr) /* FIXME: rdar://136782494 */ \
```
Ditto.  Fix this radar too.

dhecht replied:
Okay.

MenloDorian commented on Source/JavaScriptCore/heap/MarkedBlock.cpp:
```diff
@@ -513,156 +553,105 @@ void MarkedBlock::Handle::sweep(FreeList* freeList)
+    bool sawNonZero = false;
+    for (auto mem = blockStart; mem < blockStart + MarkedBlock::blockSize; mem++) {
+        // Exclude the MarkedBlock::Header::m_lock from the zero scan since taking the lock writes a non-zero value.
+        auto isMLockBytes = [blockStart](char* p) ALWAYS_INLINE_LAMBDA {
```
`isMLockBytes` implies that you're looking for the address of the `m_lock` byte when you're looking for the opposite.  Let's call this `isNotLockByte` instead.

dhecht replied:
No, I am looking for the `m_lock` bytes. Which is true when `offset` is in [`lockOffset`, `lockOffset + sizeof(m_lock)`)

MenloDorian replied:
Yeah, I completely misread the bounds check logic.

MenloDorian commented on Source/JavaScriptCore/heap/MarkedBlock.cpp:
```diff
@@ -513,156 +553,105 @@ void MarkedBlock::Handle::sweep(FreeList* freeList)
+            return lockOffset <= offset && offset < lockOffset + sizeof(MarkedBlock::Header::m_lock);
+        };
+        bool byteIsZero = !*mem;
+        if (byteIsZero || isMLockBytes(mem)) {
```
Shouldn't this be:
```
    if (!isNotLockByte(men))
        continue;
    else if (byteIsZero) {
```
Otherwise, it looks like you're counting all non-lock bytes as 0 regardless of their value.

dhecht replied:
Please see above. `isMLockBytes` should return true iff `offset` falls on `m_lock`.

MenloDorian reviewed (approved):
r=me

dhecht committed:
[JSC] Various improvements to the aboutToMarkSlow instrumentation
https://bugs.webkit.org/show_bug.cgi?id=283176
rdar://139971430

Reviewed by Mark Lam.

The MarkedBlock::Header::m_lock field sits at offset 24.
dumpInfoAndCrashForInvalidHandleV2() is always called with this
lock held, and so regardless of whether 'm_lock' was corrupted with
a zero, it will always be non-zero by the time the instrumentation reads it.
So that we can get a better guess at how many contiguious bytes
from the start of the MarkedBlock are corrupted with zeros, treat the bytes
occupied by m_lock as zeros when scanning for zeros.

In other words, we don't know what value these bytes were corrupted with,
so assume they were corrupted with zero so that the scan for zeros continues.

Also count the total number of zeros in the MarkedBlock (which is
the same as a page on Darwin) to get an overall view into the state
of the page, to account for the fact that there could be other
locations on the page that may have been written after the corruption
occurs.

Improve the use of WTF::setCrashLogMessage(). That function saves
only the last message, so send the same message on each invocation
but with additional known details filled in. Then, include
the source code line number so that the message can be interpreted
properly.

Record the first 8 bytes of the HeapCell rather than just the type.

Remove the code that tries to save values to registers in the middle
of the instrumentation. It doesn't work in practice since a crash
would be more likely to occur after running non-trivial code, which
will likely clobber the registers anyway.

Add some testing code for the various instrumentation paths
(disabled at compile time).

* Source/JavaScriptCore/heap/MarkedBlock.cpp:
(JSC::MarkedBlock::setupTestForDumpInfoAndCrash):
(JSC::MarkedBlock::aboutToMarkSlow):
(JSC::MarkedBlock::dumpInfoAndCrashForInvalidHandleV2):
(JSC::MarkedBlock::dumpInfoAndCrashForInvalidHandle): Deleted.
* Source/JavaScriptCore/heap/MarkedBlock.h:
(JSC::MarkedBlock::setupTestForDumpInfoAndCrash):
* Source/JavaScriptCore/runtime/OptionsList.h:

Canonical link: https://commits.webkit.org/286689@main
